# Импорт библиотек

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Загрузка данных

In [32]:
df_train = pd.read_csv("../data/train.csv")
df_train.sort_index(inplace=True)
df_train.head(5)

,index,"IC50, mM","CC50, mM",SI,MaxAbsEStateIndex,MaxEStateIndex,MinAbsEStateIndex,MinEStateIndex,qed,SPS,MolWt,HeavyAtomMolWt,ExactMolWt,NumValenceElectrons,NumRadicalElectrons,MaxPartialCharge,MinPartialCharge,MaxAbsPartialCharge,MinAbsPartialCharge,FpDensityMorgan1,FpDensityMorgan2,FpDensityMorgan3,BCUT2D_MWHI,BCUT2D_MWLOW,BCUT2D_CHGHI,BCUT2D_CHGLO,BCUT2D_LOGPHI,BCUT2D_LOGPLOW,BCUT2D_MRHI,BCUT2D_MRLOW,AvgIpc,BalabanJ,BertzCT,Chi0,Chi0n,Chi0v,Chi1,Chi1n,Chi1v,Chi2n,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HallKierAlpha,Ipc,Kappa1,Kappa2,Kappa3,LabuteASA,PEOE_VSA1,PEOE_VSA10,PEOE_VSA11,PEOE_VSA12,PEOE_VSA13,PEOE_VSA14,PEOE_VSA2,PEOE_VSA3,PEOE_VSA4,PEOE_VSA5,PEOE_VSA6,PEOE_VSA7,PEOE_VSA8,PEOE_VSA9,SMR_VSA1,SMR_VSA10,SMR_VSA2,SMR_VSA3,SMR_VSA4,SMR_VSA5,SMR_VSA6,SMR_VSA7,SMR_VSA8,SMR_VSA9,SlogP_VSA1,SlogP_VSA10,SlogP_VSA11,SlogP_VSA12,SlogP_VSA2,SlogP_VSA3,SlogP_VSA4,SlogP_VSA5,SlogP_VSA6,SlogP_VSA7,SlogP_VSA8,SlogP_VSA9,TPSA,EState_VSA1,EState_VSA10,EState_VSA11,EState_VSA2,EState_VSA3,EState_VSA4,EState_VSA5,EState_VSA6,EState_VSA7,EState_VSA8,EState_VSA9,VSA_EState1,VSA_EState10,VSA_EState2,VSA_EState3,VSA_EState4,VSA_EState5,VSA_EState6,VSA_EState7,VSA_EState8,VSA_EState9,FractionCSP3,HeavyAtomCount,NHOHCount,NOCount,NumAliphaticCarbocycles,NumAliphaticHeterocycles,NumAliphaticRings,NumAromaticCarbocycles,NumAromaticHeterocycles,NumAromaticRings,NumHAcceptors,NumHDonors,NumHeteroatoms,NumRotatableBonds,NumSaturatedCarbocycles,NumSaturatedHeterocycles,NumSaturatedRings,RingCount,MolLogP,MolMR,fr_Al_COO,fr_Al_OH,fr_Al_OH_noTert,fr_ArN,fr_Ar_COO,fr_Ar_N,fr_Ar_NH,fr_Ar_OH,fr_COO,fr_COO2,fr_C_O,fr_C_O_noCOO,fr_C_S,fr_HOCCN,fr_Imine,fr_NH0,fr_NH1,fr_NH2,fr_N_O,fr_Ndealkylation1,fr_Ndealkylation2,fr_Nhpyrrole,fr_SH,fr_aldehyde,fr_alkyl_carbamate,fr_alkyl_halide,fr_allylic_oxid,fr_amide,fr_amidine,fr_aniline,fr_aryl_methyl,fr_azide,fr_azo,fr_barbitur,fr_benzene,fr_benzodiazepine,fr_bicyclic,fr_diazo,fr_dihydropyridine,fr_epoxide,fr_ester,fr_ether,fr_furan,fr_guanido,fr_halogen,fr_hdrzine,fr_hdrzone,fr_imidazole,fr_imide,fr_isocyan,fr_isothiocyan,fr_ketone,fr_ketone_Topliss,fr_lactam,fr_lactone,fr_methoxy,fr_morpholine,fr_nitrile,fr_nitro,fr_nitro_arom,fr_nitro_arom_nonortho,fr_nitroso,fr_oxazole,fr_oxime,fr_para_hydroxylation,fr_phenol,fr_phenol_noOrthoHbond,fr_phos_acid,fr_phos_ester,fr_piperdine,fr_piperzine,fr_priamide,fr_prisulfonamd,fr_pyridine,fr_quatN,fr_sulfide,fr_sulfonamd,fr_sulfone,fr_term_acetylene,fr_tetrazole,fr_thiazole,fr_thiocyan,fr_thiophene,fr_unbrch_alkane,fr_urea
0,0,102.414420,95.757483,0.935000,5.466584,5.466584,0.719259,0.719259,0.681165,18.307692,195.287,182.183,195.071785,70,0,0.119177,-0.360247,0.360247,0.119177,1.230769,2.076923,2.923077,32.166506,10.291948,2.082825,-2.203512,2.178787,-2.230562,7.988993,0.155539,2.471240,2.063714,244.361475,8.932875,7.478889,8.295385,6.466326,4.421446,5.576147,2.950079,4.024775,1.987598,2.937188,1.292489,2.101284,-0.67,1.789178e+03,8.907639,4.533756,2.386264,82.881829,9.636773,6.730817,0.000000,0.000000,0.000000,0.000000,0.000,0.000000,0.000000,11.761885,18.199101,12.132734,11.440240,12.483670,4.736863,17.449271,0.0,0.000000,0.000000,0.000000,29.867250,30.331835,0,0.000000,4.899910,5.687386,0.000000,11.761885,24.967341,4.736863,0.000000,0.000000,30.331835,0.000000,0.000000,0,12.47,0.000000,0.000000,0.0,0.000000,0.000000,24.967341,5.687386,0.000000,17.828252,29.165378,4.736863,5.466584,1.919608,2.247685,0.000000,1.247130,2.132289,10.398397,0.000000,1.588307,0.000000,0.400000,13,0,2,0,1,1,1,0,1,3,0,3,1,0,1,1,2,2.1714,57.0650,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0
1,1,0.044333,8.401080,189.500000,11.492712,11.492712,0.012350,-3.798024,0.769122,27.652174,360.907,335.707,360.127441,130,0,0.237676,-0.393087,0.393087,0.237676,1.304348,1.956522,2.521739,35.495692,9.631497,2.400142,-2.449813,2.456783,-2.415599,7.887726,-0.003171,2.314947,2.157249,690.565805,17.46

In [33]:
df_test = pd.read_csv("../data/test.csv")
df_test.sort_index(inplace=True)
df_test.head(5)

,index,MaxAbsEStateIndex,MaxEStateIndex,MinAbsEStateIndex,MinEStateIndex,qed,SPS,MolWt,HeavyAtomMolWt,ExactMolWt,NumValenceElectrons,NumRadicalElectrons,MaxPartialCharge,MinPartialCharge,MaxAbsPartialCharge,MinAbsPartialCharge,FpDensityMorgan1,FpDensityMorgan2,FpDensityMorgan3,BCUT2D_MWHI,BCUT2D_MWLOW,BCUT2D_CHGHI,BCUT2D_CHGLO,BCUT2D_LOGPHI,BCUT2D_LOGPLOW,BCUT2D_MRHI,BCUT2D_MRLOW,AvgIpc,BalabanJ,BertzCT,Chi0,Chi0n,Chi0v,Chi1,Chi1n,Chi1v,Chi2n,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HallKierAlpha,Ipc,Kappa1,Kappa2,Kappa3,LabuteASA,PEOE_VSA1,PEOE_VSA10,PEOE_VSA11,PEOE_VSA12,PEOE_VSA13,PEOE_VSA14,PEOE_VSA2,PEOE_VSA3,PEOE_VSA4,PEOE_VSA5,PEOE_VSA6,PEOE_VSA7,PEOE_VSA8,PEOE_VSA9,SMR_VSA1,SMR_VSA10,SMR_VSA2,SMR_VSA3,SMR_VSA4,SMR_VSA5,SMR_VSA6,SMR_VSA7,SMR_VSA8,SMR_VSA9,SlogP_VSA1,SlogP_VSA10,SlogP_VSA11,SlogP_VSA12,SlogP_VSA2,SlogP_VSA3,SlogP_VSA4,SlogP_VSA5,SlogP_VSA6,SlogP_VSA7,SlogP_VSA8,SlogP_VSA9,TPSA,EState_VSA1,EState_VSA10,EState_VSA11,EState_VSA2,EState_VSA3,EState_VSA4,EState_VSA5,EState_VSA6,EState_VSA7,EState_VSA8,EState_VSA9,VSA_EState1,VSA_EState10,VSA_EState2,VSA_EState3,VSA_EState4,VSA_EState5,VSA_EState6,VSA_EState7,VSA_EState8,VSA_EState9,FractionCSP3,HeavyAtomCount,NHOHCount,NOCount,NumAliphaticCarbocycles,NumAliphaticHeterocycles,NumAliphaticRings,NumAromaticCarbocycles,NumAromaticHeterocycles,NumAromaticRings,NumHAcceptors,NumHDonors,NumHeteroatoms,NumRotatableBonds,NumSaturatedCarbocycles,NumSaturatedHeterocycles,NumSaturatedRings,RingCount,MolLogP,MolMR,fr_Al_COO,fr_Al_OH,fr_Al_OH_noTert,fr_ArN,fr_Ar_COO,fr_Ar_N,fr_Ar_NH,fr_Ar_OH,fr_COO,fr_COO2,fr_C_O,fr_C_O_noCOO,fr_C_S,fr_HOCCN,fr_Imine,fr_NH0,fr_NH1,fr_NH2,fr_N_O,fr_Ndealkylation1,fr_Ndealkylation2,fr_Nhpyrrole,fr_SH,fr_aldehyde,fr_alkyl_carbamate,fr_alkyl_halide,fr_allylic_oxid,fr_amide,fr_amidine,fr_aniline,fr_aryl_methyl,fr_azide,fr_azo,fr_barbitur,fr_benzene,fr_benzodiazepine,fr_bicyclic,fr_diazo,fr_dihydropyridine,fr_epoxide,fr_ester,fr_ether,fr_furan,fr_guanido,fr_halogen,fr_hdrzine,fr_hdrzone,fr_imidazole,fr_imide,fr_isocyan,fr_isothiocyan,fr_ketone,fr_ketone_Topliss,fr_lactam,fr_lactone,fr_methoxy,fr_morpholine,fr_nitrile,fr_nitro,fr_nitro_arom,fr_nitro_arom_nonortho,fr_nitroso,fr_oxazole,fr_oxime,fr_para_hydroxylation,fr_phenol,fr_phenol_noOrthoHbond,fr_phos_acid,fr_phos_ester,fr_piperdine,fr_piperzine,fr_priamide,fr_prisulfonamd,fr_pyridine,fr_quatN,fr_sulfide,fr_sulfonamd,fr_sulfone,fr_term_acetylene,fr_tetrazole,fr_thiazole,fr_thiocyan,fr_thiophene,fr_unbrch_alkane,fr_urea
0,0,13.761882,13.761882,0.121946,-0.962625,0.770057,30.580645,450.541,432.397,450.070799,156,0,0.307795,-0.480983,0.480983,0.307795,1.129032,1.967742,2.741935,32.166547,9.704289,2.608086,-2.425359,2.544663,-2.545620,8.056645,-0.147725,3.494848,1.531269,1171.998965,21.275292,16.250243,17.883236,15.058551,10.093357,11.852663,8.159777,10.171294,6.660754,8.495535,5.422732,7.148420,-2.84,2.734932e+07,18.891356,6.831720,2.610159,186.392155,10.006437,6.165776,0.000000,5.907180,5.907180,5.969305,19.283521,0.000000,0.000000,23.098671,30.351141,33.892353,16.712959,29.631406,19.490139,46.569722,0.0,4.89991,23.671624,12.710532,4.899910,74.684093,0,0.000000,4.899910,5.687386,0.000000,23.098671,33.955877,16.133831,23.671624,15.235136,64.243494,0.0,0.000000,0,77.92,23.970893,19.490139,0.0,23.650172,17.795594,9.782284,11.761885,39.335707,41.139256,0.000000,0.000000,0.000000,3.080138,45.049341,13.952413,1.061592,-3.488287,11.053051,3.404616,0.387137,0.0,0.260870,31,1,6,1,3,4,1,1,2,5,1,8,3,0,1,1,6,3.78420,118.3183,1,0,0,0,0,0,0,0,1,1,3,2,0,0,0,2,0,0,0,0,0,0,0,0,0,0,2,2,0,1,0,0,0,0,1,0,6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0
1,1,13.224489,13.224489,0.066132,-1.801871,0.278628,25.687500,448.380,428.220,448.100561,170,0,0.238633,-0.507966,0.507966,0.238633,0.906250,1.593750,2.250000,16.707906,9.991346,2.450224,-2.390353,2.369524,-2.597794,5.878314,-0.277225,2.752143,1.885708,1185.055548,23.153972,16.321448,16.321448,15.205616,9.502898,9.502898,7.289468

Удалим ненужный признак Index

In [44]:
df_train = df_train.drop(columns="index")
df_test = df_test.drop(columns="index")

## Анализ данных

### Первичный осмотр 

In [55]:
df_train.shape  


(734, 213)

In [17]:
df_test.shape

(250, 211)

Обучающий датасет представлен 751 строкой и 214 признаками Тестовый датасет представлен 250 строками, 211 признаками. 3 дополнительных признака в обучающем набором данных обусловлены наличием трех целевых переменных. Соотношение строк train/test составляет примерно 3:1.





In [18]:
df_train.dtypes.value_counts()

int64      107
float64    107
Name: count, dtype: int64

In [19]:
df_test.dtypes.value_counts()

int64      107
float64    104
Name: count, dtype: int64

Датасеты полностью состоят из числовых данных, что упрощает их использование для большинства моделей машинного обучения без необходимости дополнительного кодирования категориальных переменных. 

### Анализ пропусков 

In [22]:
df_train.isna().sum().sum()

np.int64(24)

In [23]:
for col in df_train.columns[df_train.isna().any()]:
    missing_rows = df_train[df_train[col].isna()].index.tolist()
    print(f"Колонка '{col}': пропуски в строках {missing_rows}")

Колонка 'MaxPartialCharge': пропуски в строках [331, 500]
Колонка 'MinPartialCharge': пропуски в строках [331, 500]
Колонка 'MaxAbsPartialCharge': пропуски в строках [331, 500]
Колонка 'MinAbsPartialCharge': пропуски в строках [331, 500]
Колонка 'BCUT2D_MWHI': пропуски в строках [331, 500]
Колонка 'BCUT2D_MWLOW': пропуски в строках [331, 500]
Колонка 'BCUT2D_CHGHI': пропуски в строках [331, 500]
Колонка 'BCUT2D_CHGLO': пропуски в строках [331, 500]
Колонка 'BCUT2D_LOGPHI': пропуски в строках [331, 500]
Колонка 'BCUT2D_LOGPLOW': пропуски в строках [331, 500]
Колонка 'BCUT2D_MRHI': пропуски в строках [331, 500]
Колонка 'BCUT2D_MRLOW': пропуски в строках [331, 500]


In [20]:
df_test.isna().sum().sum()

np.int64(12)

In [21]:
for col in df_test.columns[df_test.isna().any()]:
    missing_rows = df_test[df_test[col].isna()].index.tolist()
    print(f"Колонка '{col}': пропуски в строках {missing_rows}")

Колонка 'MaxPartialCharge': пропуски в строках [193]
Колонка 'MinPartialCharge': пропуски в строках [193]
Колонка 'MaxAbsPartialCharge': пропуски в строках [193]
Колонка 'MinAbsPartialCharge': пропуски в строках [193]
Колонка 'BCUT2D_MWHI': пропуски в строках [193]
Колонка 'BCUT2D_MWLOW': пропуски в строках [193]
Колонка 'BCUT2D_CHGHI': пропуски в строках [193]
Колонка 'BCUT2D_CHGLO': пропуски в строках [193]
Колонка 'BCUT2D_LOGPHI': пропуски в строках [193]
Колонка 'BCUT2D_LOGPLOW': пропуски в строках [193]
Колонка 'BCUT2D_MRHI': пропуски в строках [193]
Колонка 'BCUT2D_MRLOW': пропуски в строках [193]


In [35]:
df_train = df_train.drop([331, 500]).reset_index(drop=True)
df_train.shape  


(749, 214)

После анализа пропусков было выявлено, что отсутствующие значения присутствуют только в небольшом числе объектов: строки 331 и 500 в обучающей выборке и строка 193 в тестовой выборке. Пропуски наблюдаются исключительно в дескрипторах, связанных с расчетом электронных характеристик и BCUT-признаков (MaxPartialCharge, MinPartialCharge, BCUT2D_* и др.), что указывает не на случайную потерю данных, а на невозможность корректного вычисления данных дескрипторов для отдельных молекул средствами RDKit.

Поскольку количество проблемных объектов в обучающей выборке крайне мало (2 из 751, менее 0.3%), их удаление является оправданным и позволяет избежать внесения шума в модель. Использование стандартной импутации (mean/median) для подобных физико-химических дескрипторов нежелательно, так как это может привести к появлению химически некорректных значений.

Для тестовой выборки удаление объектов невозможно, поэтому пропуски будут оставлены без изменений. 

### Проверка константных признаков  

In [39]:
constant_cols_train = [
    c for c in df_train.columns
    if df_train[c].nunique() == 1
]

len(constant_cols_train)

18

In [40]:
print("Constant columns train:\n")

for i, col in enumerate(constant_cols_train, 1):
    print(f"{i:>2}. {col}")

Constant columns train:

 1. NumRadicalElectrons
 2. SMR_VSA8
 3. SlogP_VSA9
 4. fr_N_O
 5. fr_SH
 6. fr_azide
 7. fr_barbitur
 8. fr_benzodiazepine
 9. fr_diazo
10. fr_dihydropyridine
11. fr_isocyan
12. fr_isothiocyan
13. fr_lactam
14. fr_nitroso
15. fr_phos_acid
16. fr_phos_ester
17. fr_prisulfonamd
18. fr_thiocyan


In [41]:
constant_cols_test = [
    c for c in df_test.columns
    if df_test[c].nunique() == 1
]

len(constant_cols_test)

22

In [42]:
print("Constant columns test:\n")

for i, col in enumerate(constant_cols_test, 1):
    print(f"{i:>2}. {col}")

Constant columns test:

 1. NumRadicalElectrons
 2. SMR_VSA8
 3. SlogP_VSA9
 4. fr_Ar_COO
 5. fr_HOCCN
 6. fr_N_O
 7. fr_SH
 8. fr_azide
 9. fr_barbitur
10. fr_benzodiazepine
11. fr_diazo
12. fr_dihydropyridine
13. fr_isocyan
14. fr_isothiocyan
15. fr_lactam
16. fr_nitroso
17. fr_phos_acid
18. fr_phos_ester
19. fr_prisulfonamd
20. fr_term_acetylene
21. fr_tetrazole
22. fr_thiocyan


В ходе EDA были выявлены константные признаки, принимающие одинаковое значение для всех объектов. Большинство из них относятся к редким функциональным группам (fr_*), отсутствующим в рассматриваемом наборе молекул. Также константными оказались некоторые специализированные дескрипторы (NumRadicalElectrons, SMR_VSA8, SlogP_VSA9).

Поскольку такие признаки не содержат вариативности и не способны разделять объекты, они не несут полезной информации для модели и подлежат удалению.

Удаление выполняется на основе обучающей выборки, после чего одинаковый набор признаков сохраняется как для train, так и для test выборки.

In [ ]:
#X_train = X_train.drop(columns=constant_cols)
#X_test = X_test.drop(columns=constant_cols)

### Проверка на дубликаты строк (полные)

In [45]:
df_train.duplicated().any()

np.True_

In [49]:
duplicates = df_train[
    df_train.duplicated(keep=False)
]

print(f"Количество полных дубликатов: {len(duplicates)}")

duplicates

Количество полных дубликатов: 28


,"IC50, mM","CC50, mM",SI,MaxAbsEStateIndex,MaxEStateIndex,MinAbsEStateIndex,MinEStateIndex,qed,SPS,MolWt,HeavyAtomMolWt,ExactMolWt,NumValenceElectrons,NumRadicalElectrons,MaxPartialCharge,MinPartialCharge,MaxAbsPartialCharge,MinAbsPartialCharge,FpDensityMorgan1,FpDensityMorgan2,FpDensityMorgan3,BCUT2D_MWHI,BCUT2D_MWLOW,BCUT2D_CHGHI,BCUT2D_CHGLO,BCUT2D_LOGPHI,BCUT2D_LOGPLOW,BCUT2D_MRHI,BCUT2D_MRLOW,AvgIpc,BalabanJ,BertzCT,Chi0,Chi0n,Chi0v,Chi1,Chi1n,Chi1v,Chi2n,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HallKierAlpha,Ipc,Kappa1,Kappa2,Kappa3,LabuteASA,PEOE_VSA1,PEOE_VSA10,PEOE_VSA11,PEOE_VSA12,PEOE_VSA13,PEOE_VSA14,PEOE_VSA2,PEOE_VSA3,PEOE_VSA4,PEOE_VSA5,PEOE_VSA6,PEOE_VSA7,PEOE_VSA8,PEOE_VSA9,SMR_VSA1,SMR_VSA10,SMR_VSA2,SMR_VSA3,SMR_VSA4,SMR_VSA5,SMR_VSA6,SMR_VSA7,SMR_VSA8,SMR_VSA9,SlogP_VSA1,SlogP_VSA10,SlogP_VSA11,SlogP_VSA12,SlogP_VSA2,SlogP_VSA3,SlogP_VSA4,SlogP_VSA5,SlogP_VSA6,SlogP_VSA7,SlogP_VSA8,SlogP_VSA9,TPSA,EState_VSA1,EState_VSA10,EState_VSA11,EState_VSA2,EState_VSA3,EState_VSA4,EState_VSA5,EState_VSA6,EState_VSA7,EState_VSA8,EState_VSA9,VSA_EState1,VSA_EState10,VSA_EState2,VSA_EState3,VSA_EState4,VSA_EState5,VSA_EState6,VSA_EState7,VSA_EState8,VSA_EState9,FractionCSP3,HeavyAtomCount,NHOHCount,NOCount,NumAliphaticCarbocycles,NumAliphaticHeterocycles,NumAliphaticRings,NumAromaticCarbocycles,NumAromaticHeterocycles,NumAromaticRings,NumHAcceptors,NumHDonors,NumHeteroatoms,NumRotatableBonds,NumSaturatedCarbocycles,NumSaturatedHeterocycles,NumSaturatedRings,RingCount,MolLogP,MolMR,fr_Al_COO,fr_Al_OH,fr_Al_OH_noTert,fr_ArN,fr_Ar_COO,fr_Ar_N,fr_Ar_NH,fr_Ar_OH,fr_COO,fr_COO2,fr_C_O,fr_C_O_noCOO,fr_C_S,fr_HOCCN,fr_Imine,fr_NH0,fr_NH1,fr_NH2,fr_N_O,fr_Ndealkylation1,fr_Ndealkylation2,fr_Nhpyrrole,fr_SH,fr_aldehyde,fr_alkyl_carbamate,fr_alkyl_halide,fr_allylic_oxid,fr_amide,fr_amidine,fr_aniline,fr_aryl_methyl,fr_azide,fr_azo,fr_barbitur,fr_benzene,fr_benzodiazepine,fr_bicyclic,fr_diazo,fr_dihydropyridine,fr_epoxide,fr_ester,fr_ether,fr_furan,fr_guanido,fr_halogen,fr_hdrzine,fr_hdrzone,fr_imidazole,fr_imide,fr_isocyan,fr_isothiocyan,fr_ketone,fr_ketone_Topliss,fr_lactam,fr_lactone,fr_methoxy,fr_morpholine,fr_nitrile,fr_nitro,fr_nitro_arom,fr_nitro_arom_nonortho,fr_nitroso,fr_oxazole,fr_oxime,fr_para_hydroxylation,fr_phenol,fr_phenol_noOrthoHbond,fr_phos_acid,fr_phos_ester,fr_piperdine,fr_piperzine,fr_priamide,fr_prisulfonamd,fr_pyridine,fr_quatN,fr_sulfide,fr_sulfonamd,fr_sulfone,fr_term_acetylene,fr_tetrazole,fr_thiazole,fr_thiocyan,fr_thiophene,fr_unbrch_alkane,fr_urea
57,1309.128993,1309.128993,1.000000,12.406138,12.406138,0.004468,-4.546029,0.816512,11.687500,229.157,223.109,229.035063,84,0,0.416186,-0.462813,0.462813,0.416186,1.250000,2.000000,2.562500,19.413174,10.170213,2.311326,-2.063168,2.297411,-2.131319,5.522838,-0.137222,2.675210,2.574573,539.932052,11.759149,7.837141,7.837141,7.465345,4.320940,4.320940,3.122003,3.122003,1.958266,1.958266,1.271509,1.271509,-1.98,4919.711652,10.534893,3.549844,1.999387,87.525545,9.401129,5.760247,0.000000,5.559267,0.0,6.176299,4.794537,0.000000,13.171245,0.0,0.000000,18.199101,6.066367,17.520542,17.588396,0.000000,0.0,4.983979,0.000000,6.176299,0.000000,46.445887,0,11.454175,5.559267,13.171245,0.000000,0.000000,4.983979,6.176299,0.000000,5.563451,39.740319,0.0,11.454175,0,46.00,17.299017,17.965782,0.0,11.454175,6.066367,6.066367,18.395897,0.000000,0.000000,4.983979,4.417151,42.123134,0.000000,13.328619,0.000000,-1.811389,0.183519,4.325409,-3.232625,0.000000,0.000000,0.100000,16,1,3,0,0,0,0,2,2,2,1,6,1,0,0,0,2,2.6537,49.7667,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0
66,100.001115,68.000312,0.679996,6.598611,6.598611,0.196181,0.196181,0.607378,57.000000,179.307,158.139,179.167400,74,0,0.021356,-0.324580,0.324580,0.021356,1.000000,1.538462,2.000000,14.849076,9.431663,2.587019,-2.641307,2.622094,-2.660970,5.131443,-0.105652,1.985901,2.031302,229.509032,9.267585,8.844935,

In [50]:
df_train = df_train.drop_duplicates().reset_index(drop=True)
df_train.duplicated().any()

np.False_

In [ ]:
df_test.duplicated().any()


np.True_

In [51]:
duplicates = df_test[
    df_test.duplicated(keep=False)
]

print(f"Количество полных дубликатов: {len(duplicates)}")

duplicates

Количество полных дубликатов: 43


,MaxAbsEStateIndex,MaxEStateIndex,MinAbsEStateIndex,MinEStateIndex,qed,SPS,MolWt,HeavyAtomMolWt,ExactMolWt,NumValenceElectrons,NumRadicalElectrons,MaxPartialCharge,MinPartialCharge,MaxAbsPartialCharge,MinAbsPartialCharge,FpDensityMorgan1,FpDensityMorgan2,FpDensityMorgan3,BCUT2D_MWHI,BCUT2D_MWLOW,BCUT2D_CHGHI,BCUT2D_CHGLO,BCUT2D_LOGPHI,BCUT2D_LOGPLOW,BCUT2D_MRHI,BCUT2D_MRLOW,AvgIpc,BalabanJ,BertzCT,Chi0,Chi0n,Chi0v,Chi1,Chi1n,Chi1v,Chi2n,Chi2v,Chi3n,Chi3v,Chi4n,Chi4v,HallKierAlpha,Ipc,Kappa1,Kappa2,Kappa3,LabuteASA,PEOE_VSA1,PEOE_VSA10,PEOE_VSA11,PEOE_VSA12,PEOE_VSA13,PEOE_VSA14,PEOE_VSA2,PEOE_VSA3,PEOE_VSA4,PEOE_VSA5,PEOE_VSA6,PEOE_VSA7,PEOE_VSA8,PEOE_VSA9,SMR_VSA1,SMR_VSA10,SMR_VSA2,SMR_VSA3,SMR_VSA4,SMR_VSA5,SMR_VSA6,SMR_VSA7,SMR_VSA8,SMR_VSA9,SlogP_VSA1,SlogP_VSA10,SlogP_VSA11,SlogP_VSA12,SlogP_VSA2,SlogP_VSA3,SlogP_VSA4,SlogP_VSA5,SlogP_VSA6,SlogP_VSA7,SlogP_VSA8,SlogP_VSA9,TPSA,EState_VSA1,EState_VSA10,EState_VSA11,EState_VSA2,EState_VSA3,EState_VSA4,EState_VSA5,EState_VSA6,EState_VSA7,EState_VSA8,EState_VSA9,VSA_EState1,VSA_EState10,VSA_EState2,VSA_EState3,VSA_EState4,VSA_EState5,VSA_EState6,VSA_EState7,VSA_EState8,VSA_EState9,FractionCSP3,HeavyAtomCount,NHOHCount,NOCount,NumAliphaticCarbocycles,NumAliphaticHeterocycles,NumAliphaticRings,NumAromaticCarbocycles,NumAromaticHeterocycles,NumAromaticRings,NumHAcceptors,NumHDonors,NumHeteroatoms,NumRotatableBonds,NumSaturatedCarbocycles,NumSaturatedHeterocycles,NumSaturatedRings,RingCount,MolLogP,MolMR,fr_Al_COO,fr_Al_OH,fr_Al_OH_noTert,fr_ArN,fr_Ar_COO,fr_Ar_N,fr_Ar_NH,fr_Ar_OH,fr_COO,fr_COO2,fr_C_O,fr_C_O_noCOO,fr_C_S,fr_HOCCN,fr_Imine,fr_NH0,fr_NH1,fr_NH2,fr_N_O,fr_Ndealkylation1,fr_Ndealkylation2,fr_Nhpyrrole,fr_SH,fr_aldehyde,fr_alkyl_carbamate,fr_alkyl_halide,fr_allylic_oxid,fr_amide,fr_amidine,fr_aniline,fr_aryl_methyl,fr_azide,fr_azo,fr_barbitur,fr_benzene,fr_benzodiazepine,fr_bicyclic,fr_diazo,fr_dihydropyridine,fr_epoxide,fr_ester,fr_ether,fr_furan,fr_guanido,fr_halogen,fr_hdrzine,fr_hdrzone,fr_imidazole,fr_imide,fr_isocyan,fr_isothiocyan,fr_ketone,fr_ketone_Topliss,fr_lactam,fr_lactone,fr_methoxy,fr_morpholine,fr_nitrile,fr_nitro,fr_nitro_arom,fr_nitro_arom_nonortho,fr_nitroso,fr_oxazole,fr_oxime,fr_para_hydroxylation,fr_phenol,fr_phenol_noOrthoHbond,fr_phos_acid,fr_phos_ester,fr_piperdine,fr_piperzine,fr_priamide,fr_prisulfonamd,fr_pyridine,fr_quatN,fr_sulfide,fr_sulfonamd,fr_sulfone,fr_term_acetylene,fr_tetrazole,fr_thiazole,fr_thiocyan,fr_thiophene,fr_unbrch_alkane,fr_urea
1,13.224489,13.224489,0.066132,-1.801871,0.278628,25.687500,448.380,428.220,448.100561,170,0,0.238633,-0.507966,0.507966,0.238633,0.906250,1.593750,2.250000,16.707906,9.991346,2.450224,-2.390353,2.369524,-2.597794,5.878314,-0.277225,2.752143,1.885708,1185.055548,23.153972,16.321448,16.321448,15.205616,9.502898,9.502898,7.289468,7.289468,5.129606,5.129606,3.459029,3.459029,-3.35,1.303530e+07,21.865906,8.297053,3.720411,179.107049,49.636569,52.633645,5.760247,17.468329,0.000000,0.000000,4.794537,0.000000,0.000000,0.000000,0.000000,24.265468,17.696186,6.606882,49.636569,10.969244,0.0,0.000000,0.000000,30.705892,6.606882,46.621530,0,34.321746,10.165653,0.000000,22.998047,0.000000,73.058466,4.736863,0.000000,0.000000,45.609891,0.0,22.292943,0,190.28,54.240588,40.540229,0.0,33.791967,0.000000,12.132734,24.265468,0.000000,0.000000,0.000000,13.890877,16.581293,0.000000,13.224489,68.782575,-0.801170,-1.732766,7.479972,-8.161395,-0.706331,0.000000,0.285714,32,7,11,0,1,1,2,1,3,11,7,11,4,0,1,1,4,-0.24450,107.3096,0,4,4,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,1,0,0,0,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
2,6.191528,6.191528,0.445278,0.445278,0.657472,55.384615,179.307,158.139,179.167400,74,0,0.006729,-0.327440,0.327440,0.006729,1.000000,1.384615,1.692308,14.717158,9.477838,2.516654,-2.580818,2.603245,-2.529164,5.037930,-0.063729,2.001545,1.911265,183.880154,9.052042,8.629392,8.629392,6.178270,5.934254,5.934254,6.068603,6.068603,5.

В ходе анализа данных были обнаружены полные дубликаты строк как в обучающей, так и в тестовой выборке. В обучающей выборке найдено 28 полностью совпадающих объектов, а в тестовой — 43.

Для тестовой выборки дубликаты были сохранены, поскольку при формировании итогового submission необходимо получить предсказание для каждого исходного объекта. Удаление строк из test-признаков может привести к несоответствию формата ответа требованиям соревнования. В обучающей выборке они были удалены.

### Выделение групп 

In [54]:
targets = ["IC50, mM", "CC50, mM", "SI"]

# ============================================================================
# ГРУППА 1: Общие молекулярные свойства
# ============================================================================
'''
Описывают базовые физико-химические характеристики молекулы:
массу, размер, липофильность, полярность, площадь поверхности,
лекарствоподобность и общую структурную сложность.
'''

group_general_properties = [
    'qed',
    'SPS',
    'MolWt',
    'HeavyAtomMolWt',
    'ExactMolWt',
    'NumValenceElectrons',
    'NumRadicalElectrons',
    'MolLogP',
    'MolMR',
    'TPSA',
    'LabuteASA',
    'HeavyAtomCount',
    'FractionCSP3',

    'NHOHCount',
    'NOCount',
    'NumHAcceptors',
    'NumHDonors',
    'NumHeteroatoms',
    'NumRotatableBonds'
]

# ============================================================================
# ГРУППА 2: Электронные и зарядовые характеристики
# ============================================================================
'''
Описывают распределение электронной плотности и зарядов в молекуле.
PEOE_VSA дескрипторы характеризуют распределение поверхности
по диапазонам частичных зарядов.
'''

group_electronic_charge = [
    'MaxPartialCharge',
    'MinPartialCharge',
    'MaxAbsPartialCharge',
    'MinAbsPartialCharge',
]

group_peoe_vsa = [
    c for c in df_train.columns
    if c.startswith("PEOE_VSA")
]

# ============================================================================
# ГРУППА 3: Топологические и структурные дескрипторы
# ============================================================================
'''
Описывают форму, связность и сложность молекулярного графа.
Chi-дескрипторы характеризуют связность,
Kappa — разветвлённость и цикличность,
BalabanJ и BertzCT — топологическую сложность молекулы.
'''

group_kappa = [
    c for c in df_train.columns
    if c.startswith("Kappa")
]

group_chi = [
    c for c in df_train.columns
    if c.startswith("Chi")
]

group_ring_descriptors = [
    'RingCount',

    'NumAliphaticCarbocycles',
    'NumAliphaticHeterocycles',
    'NumAliphaticRings',

    'NumAromaticCarbocycles',
    'NumAromaticHeterocycles',
    'NumAromaticRings',

    'NumSaturatedCarbocycles',
    'NumSaturatedHeterocycles',
    'NumSaturatedRings',
]

group_topological_structural = [

    'AvgIpc',
    'BalabanJ',
    'BertzCT',
    'HallKierAlpha',
    'Ipc',

    *group_kappa,
    *group_chi,
    *group_ring_descriptors
]

# ============================================================================
# ГРУППА 4: Поверхностные дескрипторы (VSA)
# ============================================================================
'''
Описывают распределение ван-дер-ваальсовой поверхности молекулы
по различным физико-химическим свойствам.

• SlogP_VSA  — распределение липофильности
• SMR_VSA    — распределение молярной рефрактивности
• EState_VSA — распределение EState по поверхности
• VSA_EState — распределение поверхности по уровням EState
'''

group_slogp_vsa = [
    c for c in df_train.columns
    if c.startswith("SlogP_VSA")
]

group_smr_vsa = [
    c for c in df_train.columns
    if c.startswith("SMR_VSA")
]

group_estate_vsa = [
    c for c in df_train.columns
    if c.startswith("EState_VSA")
]

group_vsa_estate = [
    c for c in df_train.columns
    if c.startswith("VSA_EState")
]

group_surface_vsa = [
    *group_slogp_vsa,
    *group_smr_vsa,
    *group_estate_vsa,
    *group_vsa_estate
]

# ============================================================================
# ГРУППА 5: Функциональные группы
# ============================================================================
'''
Признаки, отражающие количество функциональных групп
и химических фрагментов в молекуле.
Используются для SAR/QSAR анализа.
'''

group_functional_groups = [
    c for c in df_train.columns
    if c.startswith("fr_")
]

# ============================================================================
# ГРУППА 6: Специфические дескрипторы
# ============================================================================
'''
Включает специализированные хемоинформатические дескрипторы:

• EState indices —
  электротопологические индексы

• BCUT2D —
  спектральные дескрипторы молекулярного графа

• FpDensityMorgan —
  плотность Morgan fingerprints
'''

group_estate_indices = [
    'MaxAbsEStateIndex',
    'MaxEStateIndex',
    'MinAbsEStateIndex',
    'MinEStateIndex',
]

group_bcut = [
    c for c in df_train.columns
    if c.startswith("BCUT2D")
]

group_morgan_density = [
    c for c in df_train.columns
    if c.startswith("FpDensityMorgan")
]

group_specific_descriptors = [
    *group_estate_indices,
    *group_bcut,
    *group_morgan_density
]

# ============================================================================
# ИТОГОВЫЙ СЛОВАРЬ ГРУПП
# ============================================================================

feature_groups = {
    "general_properties": group_general_properties,
    "electronic_charge": group_electronic_charge,
    "peoe_vsa": group_peoe_vsa,
    "topological_structural": group_topological_structural,
    "surface_vsa": group_surface_vsa,
    "functional_groups": group_functional_groups,
    "specific_descriptors": group_specific_descriptors,
}

# ============================================================================
# ПРОВЕРКА РАЗМЕРОВ ГРУПП
# ============================================================================

for group_name, cols in feature_groups.items():
    print(f"{group_name}: {len(cols)}")

general_properties: 19
electronic_charge: 4
peoe_vsa: 14
topological_structural: 30
surface_vsa: 43
functional_groups: 85
specific_descriptors: 15
